In [18]:
import pandas as pd
import numpy as np

### Extracting my CSV file

In [19]:
# Load the dataset
featureEng = pd.read_csv('clean_data_after_eda.csv', encoding= 'latin-1')

In [20]:
featureEng.head(3)

id                     channel_sales  \
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1  d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2  764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   

   cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0         0         54946                0  2013-06-15  2016-06-15   
1      4660             0                0  2009-08-21  2016-08-30   
2       544             0                0  2010-04-16  2016-04-16   

  date_modif_prod date_renewal  forecast_cons_12m  ...  \
0      2015-11-01   2015-06-23               0.00  ...   
1      2009-08-21   2015-08-31             189.95  ...   
2      2010-04-16   2015-04-17              47.96  ...   

   var_6m_price_off_peak_var  var_6m_price_peak_var  \
0                   0.000131           4.100838e-05   
1                   0.000003           1.217891e-03   
2                   0.000004           9.450150e-08   

   var_6m_price_mid_peak_var  var_6m_price_off_peak_fix  \
0                   0.000908                   2.086294   
1                   0.000000                   0.009482   
2                   0.000000                   0.000000   

   var_6m_price_peak_fix  var_6m_price_mid_peak_fix var_6m_price_off_peak  \
0              99.530517                  44.235794              2.086425   
1               0.000000                   0.000000              0.009485   
2               0.000000                   0.000000              0.000004   

   var_6m_price_peak  var_6m_price_mid_peak  churn  
0       9.953056e+01              44.236702      1  
1       1.217891e-03               0.000000      0  
2       9.450150e-08               0.000000      0  

[3 rows x 44 columns]

In [21]:
featureEng.shape

(14606, 44)

In [22]:
featureEng.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14606 entries, 0 to 14605
Data columns (total 44 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              14606 non-null  object 
 1   channel_sales                   14606 non-null  object 
 2   cons_12m                        14606 non-null  int64  
 3   cons_gas_12m                    14606 non-null  int64  
 4   cons_last_month                 14606 non-null  int64  
 5   date_activ                      14606 non-null  object 
 6   date_end                        14606 non-null  object 
 7   date_modif_prod                 14606 non-null  object 
 8   date_renewal                    14606 non-null  object 
 9   forecast_cons_12m               14606 non-null  float64
 10  forecast_cons_year              14606 non-null  int64  
 11  forecast_discount_energy        14606 non-null  float64
 12  forecast_meter_rent_12m         

In [23]:
featureEng.describe()

cons_12m  cons_gas_12m  cons_last_month  forecast_cons_12m  \
count  1.460600e+04  1.460600e+04     14606.000000       14606.000000   
mean   1.592203e+05  2.809238e+04     16090.269752        1868.614880   
std    5.734653e+05  1.629731e+05     64364.196422        2387.571531   
min    0.000000e+00  0.000000e+00         0.000000           0.000000   
25%    5.674750e+03  0.000000e+00         0.000000         494.995000   
50%    1.411550e+04  0.000000e+00       792.500000        1112.875000   
75%    4.076375e+04  0.000000e+00      3383.000000        2401.790000   
max    6.207104e+06  4.154590e+06    771203.000000       82902.830000   

       forecast_cons_year  forecast_discount_energy  forecast_meter_rent_12m  \
count        14606.000000              14606.000000             14606.000000   
mean          1399.762906                  0.966726                63.086871   
std           3247.786255                  5.108289                66.165783   
min              0.000000                  0.000000                 0.000000   
25%              0.000000                  0.000000                16.180000   
50%            314.000000                  0.000000                18.795000   
75%           1745.750000                  0.000000               131.030000   
max         175375.000000                 30.000000               599.310000   

       forecast_price_energy_off_peak  forecast_price_energy_peak  \
count                    14606.000000                14606.000000   
mean                         0.137283                    0.050491   
std                          0.024623                    0.049037   
min                          0.000000                    0.000000   
25%                          0.116340                    0.000000   
50%                          0.143166                    0.084138   
75%                          0.146348                    0.098837   
max                          0.273963                    0.195975   

       forecast_price_pow_off_peak  ...  var_6m_price_off_peak_var  \
count                 14606.000000  ...               14606.000000   
mean                     43.130056  ...                   0.000025   
std                       4.485988  ...                   0.000204   
min                       0.000000  ...                   0.000000   
25%                      40.606701  ...                   0.000003   
50%                      44.311378  ...                   0.000004   
75%                      44.311378  ...                   0.000011   
max                      59.266378  ...                   0.005543   

       var_6m_price_peak_var  var_6m_price_mid_peak_var  \
count           1.460600e+04               1.460600e+04   
mean            3.304264e-05               1.406015e-05   
std             2.231791e-04               1.273170e-04   
min             0.000000e+00               0.000000e+00   
25%             0.000000e+00               0.000000e+00   
50%             9.450150e-08               0.000000e+00   
75%             2.896760e-06               4.860000e-10   
max             4.791197e-03               2.415288e-03   

       var_6m_price_off_peak_fix  var_6m_price_peak_fix  \
count               14606.000000           14606.000000   
mean                    0.922810               1.460247   
std                    18.218286              13.760009   
min                     0.000000               0.000000   
25%                     0.000000               0.000000   
50%                     0.000000               0.000000   
75%                     0.007962               0.000000   
max                   525.988881             399.471115   

       var_6m_price_mid_peak_fix  var_6m_price_off_peak  var_6m_price_peak  \
count               14606.000000           14606.000000       1.460600e+04   
mean                    0.634413               0.922835       1.460280e+00   
std                     5.895346              18.218452       1.376010e+01   
min

In [24]:
featureEng.isnull().sum()

id                                0
channel_sales                     0
cons_12m                          0
cons_gas_12m                      0
cons_last_month                   0
date_activ                        0
date_end                          0
date_modif_prod                   0
date_renewal                      0
forecast_cons_12m                 0
forecast_cons_year                0
forecast_discount_energy          0
forecast_meter_rent_12m           0
forecast_price_energy_off_peak    0
forecast_price_energy_peak        0
forecast_price_pow_off_peak       0
has_gas                           0
imp_cons                          0
margin_gross_pow_ele              0
margin_net_pow_ele                0
nb_prod_act                       0
net_margin                        0
num_years_antig                   0
origin_up                         0
pow_max                           0
var_year_price_off_peak_var       0
var_year_price_peak_var           0
var_year_price_mid_peak_var 

## What We're Doing Here

We're **creating new features from dates** to help predict customer churn. Think of it like turning calendar dates into useful information that a computer can understand.

## Why We're Doing This

**Original Problem**: We have dates like "2020-03-15" but dates alone don't tell us much about customer behavior.

**Solution**: Convert dates into meaningful numbers that reveal patterns about when customers are likely to leave.

## What Each Feature Tells Us

Let me break down each feature in simple terms:

### **Customer Timing Patterns**
- **`activation_year`** → What year did they join? (2019, 2020, etc.)
- **`activation_month`** → What month? (January=1, December=12)
- **`activation_quarter`** → What season? (Q1, Q2, Q3, Q4)
- **`activation_day_of_week`** → What day? (Monday=0, Sunday=6)
- **`activation_is_weekend`** → Did they join on weekend? (Yes=1, No=0)

### **How Long They've Been With Us**
- **`days_since_activation`** → How many days since they became a customer?
- **`contract_duration_days`** → How long is their total contract?

### **Contract Status**
- **`days_until_contract_end`** → How many days until contract expires?
- **`contract_expired`** → Has their contract already expired? (Yes=1, No=0)

### **Customer Activity Timeline**
- **`days_since_product_modification`** → How long since they changed their plan?
- **`months_activation_to_renewal`** → How long from joining to renewal?
- **`days_modification_to_renewal`** → Time between plan change and renewal?

## What We Can Learn From This

### **Churn Risk Patterns**:
1. **New customers** (low `days_since_activation`) might be more likely to leave
2. **Contract expiration** (`contract_expired=1`) is a high-risk time
3. **Seasonal patterns** - maybe more people leave in January after holidays
4. **Weekend joiners** might behave differently than weekday joiners

### **Business Insights**:
- **Which months** have the most churn?
- **How long** do customers typically stay?
- **When** should we reach out to prevent churn?

## Simple Example

**Before**: "Customer joined on 2020-03-15"
**After**: 
- activation_year = 2020
- activation_month = 3 (March)
- days_since_activation = 1,412 days
- activation_is_weekend = 0 (Sunday)

Now the computer can see patterns like "customers who joined in March tend to stay longer" or "weekend joiners are 20% more likely to churn."

**Bottom Line**: We're turning messy dates into clean numbers that help us predict which customers might leave!

In [25]:
# Convert date columns to datetime format
date_columns = ['date_activ', 'date_end', 'date_renewal', 'date_modif_prod']

for col in date_columns:
    if col in featureEng.columns:
        featureEng[col] = pd.to_datetime(featureEng[col], errors='coerce')

# Set a reference date (current date)
# This can be adjusted based on the context of the analysis
reference_date = pd.Timestamp.now()


# 1. Activation year, month, and duration since activation
if 'date_activ' in featureEng.columns:
    featureEng['activation_year'] = featureEng['date_activ'].dt.year
    featureEng['activation_month'] = featureEng['date_activ'].dt.month
    featureEng['activation_day_of_week'] = featureEng['date_activ'].dt.dayofweek
    featureEng['days_since_activation'] = (reference_date - featureEng['date_activ']).dt.days

# 2. Days until end of contract
if 'date_end' in featureEng.columns:
    featureEng['days_until_contract_end'] = (featureEng['date_end'] - reference_date).dt.days
    # Also create a binary feature for expired contracts
    featureEng['contract_expired'] = (featureEng['date_end'] < reference_date).astype(int)

# 3. Days since last product modification
if 'date_modif_prod' in featureEng.columns:
    featureEng['days_since_product_modification'] = (reference_date - featureEng['date_modif_prod']).dt.days

# 4. Months between activation and renewal/churn
if 'date_activ' in featureEng.columns and 'date_renewal' in featureEng.columns:
    featureEng['months_activation_to_renewal'] = ((featureEng['date_renewal'] - featureEng['date_activ']).dt.days / 30.44).round(2)

if 'date_activ' in featureEng.columns and 'date_end' in featureEng.columns:
    featureEng['months_activation_to_end'] = ((featureEng['date_end'] - featureEng['date_activ']).dt.days / 30.44).round(2)

# Additional useful time-based features
if 'date_activ' in featureEng.columns:
    featureEng['activation_quarter'] = featureEng['date_activ'].dt.quarter
    featureEng['activation_is_weekend'] = (featureEng['date_activ'].dt.dayofweek >= 5).astype(int)

# Contract duration features
if 'date_activ' in featureEng.columns and 'date_end' in featureEng.columns:
    featureEng['contract_duration_days'] = (featureEng['date_end'] - featureEng['date_activ']).dt.days

# Time between modifications and renewals
if 'date_modif_prod' in featureEng.columns and 'date_renewal' in featureEng.columns:
    featureEng['days_modification_to_renewal'] = (featureEng['date_renewal'] - featureEng['date_modif_prod']).dt.days

# Display the new time-based features
time_features = [col for col in featureEng.columns if any(keyword in col.lower() for keyword in 
                ['activation', 'days_', 'months_', 'contract_', 'expired', 'weekend', 'quarter'])]

print("New time-based features created:")
for feature in time_features:
    print(f"- {feature}")

# Show sample of the new features
print("\nSample of new time-based features:")
featureEng[time_features].head()

New time-based features created:
- activation_year
- activation_month
- activation_day_of_week
- days_since_activation
- days_until_contract_end
- contract_expired
- days_since_product_modification
- months_activation_to_renewal
- months_activation_to_end
- activation_quarter
- activation_is_weekend
- contract_duration_days
- days_modification_to_renewal

Sample of new time-based features:


activation_year  activation_month  activation_day_of_week  \
0             2013                 6                       5   
1             2009                 8                       4   
2             2010                 4                       4   
3             2010                 3                       1   
4             2010                 1                       2   

   days_since_activation  days_until_contract_end  contract_expired  \
0                   4400                    -3305                 1   
1                   5794                    -3229                 1   
2                   5556                    -3365                 1   
3                   5573                    -3382                 1   
4                   5649                    -3405                 1   

   days_since_product_modification  months_activation_to_renewal  \
0                             3531                         24.24   
1                             5794                         72.31   
2                             5556                         60.02   
3                             5573                         60.02   
4                             5649                         61.79   

   months_activation_to_end  activation_quarter  activation_is_weekend  \
0                     36.01                   2                      1   
1                     84.30                   3                      0   
2                     72.01                   2                      0   
3                     72.01                   1                      0   
4                     73.75                   1                      0   

   contract_duration_days  days_modification_to_renewal  
0                    1096                          -131  
1                    2566                          2201  
2                    2192                          1827  
3                    2192                          1827  
4                    2245                          1881

## What We're Doing Here - Price Feature Engineering

We're creating **smart price features** to understand how **seasonal price changes** affect customer behavior and churn risk.

## Why This Matters

**Problem**: Energy prices change throughout the year, especially between winter (Dec-Jan) and other seasons. These price swings might make customers unhappy and leave.

**Solution**: Create features that capture these seasonal price patterns so we can predict which customers are at risk.

## Breaking Down Each Feature Type

### **1. Seasonal Price Proxies (Dec-Jan)**
- **`dec_jan_proxy_offpeak/peak/midpeak`** → Uses yearly price variations as estimates for winter price changes
- **Think of it as**: "How much do prices typically swing during winter months?"

### **2. Seasonal Differences**
- **`seasonal_price_diff_offpeak/peak/midpeak`** → Difference between yearly and 6-month price changes
- **Think of it as**: "Are recent price changes bigger or smaller than usual seasonal patterns?"

### **3. Price Ratios**
- **`offpeak_to_peak_ratio`** → How off-peak prices compare to peak prices
- **`offpeak_to_midpeak_ratio`** → How off-peak compares to mid-peak
- **Think of it as**: "How big is the price difference between different times of day?"

### **4. Winter Customer Behavior**
- **`winter_activation`** → Did customer join in winter? (Dec/Jan = 1, else = 0)
- **`winter_offpeak_interaction`** → Winter customers × price variations
- **`winter_price_sensitivity`** → How winter customers react to seasonal price differences

### **5. Price Volatility (How Unpredictable Prices Are)**
- **`price_volatility_total/avg/max`** → Different ways to measure price instability
- **Think of it as**: "How crazy and unpredictable are this customer's prices?"

### **6. Fixed vs Variable Price Differences**
- **`offpeak_fix_var_diff`** → Difference between fixed and variable pricing
- **Think of it as**: "Is this customer paying more for predictable vs. flexible pricing?"

## What We Can Learn From The Sample Data

Looking at the first few customers:

**Customer 1 (Row 0)**:
- Contract duration: 1096 days (~3 years)
- Low price volatility (0.000176 avg) → Stable prices
- Not a winter joiner (winter_activation = 0)
- **Risk Level**: Probably LOW (stable prices, long contract)

**Customer 5 (Row 4)**:
- Contract duration: 2245 days (~6 years)
- Higher price volatility (0.000007 avg)
- IS a winter joiner (winter_activation = 1) → More price sensitive
- Winter interaction shows they're affected by seasonal changes
- **Risk Level**: Possibly HIGHER (winter joiner + some volatility)

## Business Insights We Can Get

### **High Churn Risk Patterns**:
1. **High price volatility** → Customers with unpredictable bills
2. **Winter joiners** with high price sensitivity → Seasonal price shock
3. **Large seasonal differences** → Customers experiencing big price swings
4. **High peak/off-peak ratios** → Customers paying big premiums

### **Retention Strategies**:
- **Target winter joiners** with special offers during price-sensitive periods
- **Offer fixed pricing** to customers with high volatility
- **Monitor seasonal patterns** to predict when customers might leave

## Simple Summary

**We're basically asking**: 
- "How much do prices jump around for each customer?"
- "Are winter customers more likely to leave?"
- "Do big seasonal price changes scare customers away?"

**The goal**: Use these price patterns to predict which customers might churn and intervene before they leave!

In [26]:
# Load and process price data to calculate actual December-January differences
print("Loading price data and calculating actual December-January price differences...")

# Load price data
price_df = pd.read_csv('price_data.csv')
price_df["price_date"] = pd.to_datetime(price_df["price_date"], format='%Y-%m-%d')

# Extract month from price_date
price_df['month'] = price_df['price_date'].dt.month

# Filter for December and January data only
dec_jan_data = price_df[price_df['month'].isin([12, 1])].copy()

# Group by customer ID and month to get average monthly prices
monthly_prices = dec_jan_data.groupby(['id', 'month']).agg({
    'price_off_peak_var': 'mean',
    'price_peak_var': 'mean', 
    'price_mid_peak_var': 'mean',
    'price_off_peak_fix': 'mean',
    'price_peak_fix': 'mean',
    'price_mid_peak_fix': 'mean'
}).reset_index()

# Separate December and January data
dec_prices = monthly_prices[monthly_prices['month'] == 12].drop('month', axis=1)
jan_prices = monthly_prices[monthly_prices['month'] == 1].drop('month', axis=1)

# Rename columns for merging
dec_prices = dec_prices.rename(columns={
    'price_off_peak_var': 'dec_price_off_peak_var',
    'price_peak_var': 'dec_price_peak_var',
    'price_mid_peak_var': 'dec_price_mid_peak_var',
    'price_off_peak_fix': 'dec_price_off_peak_fix',
    'price_peak_fix': 'dec_price_peak_fix',
    'price_mid_peak_fix': 'dec_price_mid_peak_fix'
})

jan_prices = jan_prices.rename(columns={
    'price_off_peak_var': 'jan_price_off_peak_var',
    'price_peak_var': 'jan_price_peak_var',
    'price_mid_peak_var': 'jan_price_mid_peak_var',
    'price_off_peak_fix': 'jan_price_off_peak_fix',
    'price_peak_fix': 'jan_price_peak_fix',
    'price_mid_peak_fix': 'jan_price_mid_peak_fix'
})

# Merge December and January prices
dec_jan_diff = pd.merge(dec_prices, jan_prices, on='id', how='outer')

# Calculate December-January differences (December - January)
dec_jan_diff['dec_jan_diff_off_peak_var'] = dec_jan_diff['dec_price_off_peak_var'] - dec_jan_diff['jan_price_off_peak_var']
dec_jan_diff['dec_jan_diff_peak_var'] = dec_jan_diff['dec_price_peak_var'] - dec_jan_diff['jan_price_peak_var']
dec_jan_diff['dec_jan_diff_mid_peak_var'] = dec_jan_diff['dec_price_mid_peak_var'] - dec_jan_diff['jan_price_mid_peak_var']
dec_jan_diff['dec_jan_diff_off_peak_fix'] = dec_jan_diff['dec_price_off_peak_fix'] - dec_jan_diff['jan_price_off_peak_fix']
dec_jan_diff['dec_jan_diff_peak_fix'] = dec_jan_diff['dec_price_peak_fix'] - dec_jan_diff['jan_price_peak_fix']
dec_jan_diff['dec_jan_diff_mid_peak_fix'] = dec_jan_diff['dec_price_mid_peak_fix'] - dec_jan_diff['jan_price_mid_peak_fix']

# Calculate absolute differences (magnitude of change regardless of direction)
dec_jan_diff['dec_jan_abs_diff_off_peak_var'] = dec_jan_diff['dec_jan_diff_off_peak_var'].abs()
dec_jan_diff['dec_jan_abs_diff_peak_var'] = dec_jan_diff['dec_jan_diff_peak_var'].abs()
dec_jan_diff['dec_jan_abs_diff_mid_peak_var'] = dec_jan_diff['dec_jan_diff_mid_peak_var'].abs()

# Calculate average differences across all price types
dec_jan_diff['dec_jan_avg_diff_var'] = dec_jan_diff[['dec_jan_diff_off_peak_var', 'dec_jan_diff_peak_var', 'dec_jan_diff_mid_peak_var']].mean(axis=1)
dec_jan_diff['dec_jan_avg_diff_fix'] = dec_jan_diff[['dec_jan_diff_off_peak_fix', 'dec_jan_diff_peak_fix', 'dec_jan_diff_mid_peak_fix']].mean(axis=1)

# Select only the ID and difference columns for merging
price_features = dec_jan_diff[['id', 'dec_jan_diff_off_peak_var', 'dec_jan_diff_peak_var', 'dec_jan_diff_mid_peak_var',
                              'dec_jan_diff_off_peak_fix', 'dec_jan_diff_peak_fix', 'dec_jan_diff_mid_peak_fix',
                              'dec_jan_abs_diff_off_peak_var', 'dec_jan_abs_diff_peak_var', 'dec_jan_abs_diff_mid_peak_var',
                              'dec_jan_avg_diff_var', 'dec_jan_avg_diff_fix']]

# Merge with main dataset
featureEng = featureEng.merge(price_features, on='id', how='left')

print(f"December-January price differences calculated for {len(price_features)} customers")
print(f"Features added to main dataset. New shape: {featureEng.shape}")

# Show sample of new features
print("\nSample of actual December-January price difference features:")
actual_dec_jan_features = [col for col in featureEng.columns if 'dec_jan_diff' in col or 'dec_jan_abs' in col or 'dec_jan_avg' in col]
print(f"New actual Dec-Jan features: {actual_dec_jan_features}")
print(featureEng[actual_dec_jan_features].head())

# Show some statistics
print(f"\nDecember-January price difference statistics:")
print(f"Customers with both Dec and Jan data: {dec_jan_diff.dropna().shape[0]}")
print(f"Average off-peak variable price difference (Dec-Jan): {dec_jan_diff['dec_jan_diff_off_peak_var'].mean():.6f}")
print(f"Average absolute off-peak difference: {dec_jan_diff['dec_jan_abs_diff_off_peak_var'].mean():.6f}")

Loading price data and calculating actual December-January price differences...
December-January price differences calculated for 16096 customers
Features added to main dataset. New shape: (14606, 68)

Sample of actual December-January price difference features:
New actual Dec-Jan features: ['dec_jan_diff_off_peak_var', 'dec_jan_diff_peak_var', 'dec_jan_diff_mid_peak_var', 'dec_jan_diff_off_peak_fix', 'dec_jan_diff_peak_fix', 'dec_jan_diff_mid_peak_fix', 'dec_jan_abs_diff_off_peak_var', 'dec_jan_abs_diff_peak_var', 'dec_jan_abs_diff_mid_peak_var', 'dec_jan_avg_diff_var', 'dec_jan_avg_diff_fix']
   dec_jan_diff_off_peak_var  dec_jan_diff_peak_var  \
0                   0.020057              -0.017912   
1                  -0.003767               0.000000   
2                  -0.004670               0.000528   
3                  -0.004547               0.000000   
4                  -0.006192              -0.002302   

   dec_jan_diff_mid_peak_var  dec_jan_diff_off_peak_fix  \
0       

## What We're Doing Here - Advanced Price Analysis

This code creates **sophisticated price features** that capture **trends, momentum, and customer risk patterns**. Let me break it down in simple terms:

## **Section 8: Price Momentum Features**

### **What is "Momentum"?**
Think of momentum like a **price trend direction**:
- **Positive momentum** = Prices are going UP over time
- **Negative momentum** = Prices are going DOWN over time  
- **Zero momentum** = Prices are stable

### **How We Calculate It:**
```python
momentum = yearly_price_change - 6month_price_change
```

**Example:**
- If yearly change = +10% and 6-month change = +5%
- Momentum = +5% → Prices are **accelerating upward**
- **Risk**: Customer might get shocked by increasing bills

### **Business Logic:**
- **High positive momentum** → Bills getting more expensive → Higher churn risk
- **High negative momentum** → Bills getting cheaper → Lower churn risk
- **Volatile momentum** → Unpredictable bills → Customer frustration

---

## **Section 9: Price Spread Features**

### **What is "Price Spread"?**
The **difference between peak and off-peak prices**:
```python
spread = peak_price - offpeak_price
```

### **What This Tells Us:**
- **High spread** → Big difference between expensive and cheap hours
- **Low spread** → Prices are similar throughout the day

### **Mid-Peak Positioning:**
Shows **where mid-peak prices sit** between off-peak and peak:
- **Position = 0** → Mid-peak = Off-peak (cheap)
- **Position = 1** → Mid-peak = Peak (expensive)  
- **Position = 0.5** → Mid-peak is exactly in the middle

### **Business Logic:**
- **High spreads** → Customers with flexible usage save money, others pay more
- **Customers paying peak prices** → Higher bills → More likely to churn

---

## **Section 10: Comprehensive Risk Scores**

### **Total Price Variation Exposure:**
**How much price instability** a customer faces overall:
```python
total_exposure = sum(all_price_variations)
```

### **Price Stability Score:**
**How predictable** a customer's bills are:
```python
stability = 1 / (total_exposure + 0.001)
```
- **High score** → Stable, predictable bills
- **Low score** → Chaotic, unpredictable bills

### **Price Complexity:**
**How many price shocks** a customer experiences:
```python
complexity = count(price_variations > threshold)
```

---

## **What We Can Learn From This**

### **Customer Risk Profiling:**

**HIGH CHURN RISK Customers:**
1. **High momentum** → Bills rapidly increasing
2. **High spreads** → Paying premium prices during peak hours
3. **Low stability score** → Unpredictable monthly bills
4. **High complexity** → Multiple price shocks

**LOW CHURN RISK Customers:**
1. **Stable momentum** → Consistent pricing
2. **Low spreads** → Fair pricing throughout day
3. **High stability score** → Predictable bills
4. **Low complexity** → Simple, stable pricing

### **Business Insights:**

#### **Retention Strategies:**
- **Target high momentum customers** → Offer fixed-rate plans
- **Help high spread customers** → Educate on off-peak usage
- **Support low stability customers** → Provide bill smoothing options

#### **Pricing Strategies:**
- **Monitor momentum trends** → Predict customer dissatisfaction
- **Reduce price spreads** → Make pricing fairer
- **Simplify complex pricing** → Improve customer experience

### **Real-World Example:**

**Customer A:**
- High positive momentum → Bills increasing 20% year-over-year
- High price spread → Pays 300% more during peak hours
- Low stability score → Bill varies wildly month to month
- **Prediction**: **VERY HIGH** churn risk

**Customer B:**
- Zero momentum → Stable pricing
- Low price spread → Fair pricing
- High stability score → Predictable bills
- **Prediction**: **LOW** churn risk

### **The Big Picture:**

This code essentially asks:
1. **"Are this customer's bills getting more expensive?"** (Momentum)
2. **"Is this customer paying unfair prices?"** (Spreads)  
3. **"How chaotic is this customer's pricing experience?"** (Stability & Complexity)

**The goal:** Identify customers who are frustrated with their pricing experience **before** they churn, so we can intervene with better offers or pricing plans.

**Bottom line:** We're turning complex price data into simple risk scores that help us predict and prevent customer churn!

In [27]:
# Engineer comprehensive price-related features
print("Creating comprehensive price-related features...")

# 1. Mean price changes across all types for 6-month period
var_6m_columns = ['var_6m_price_off_peak_var', 'var_6m_price_peak_var', 'var_6m_price_mid_peak_var']
fix_6m_columns = ['var_6m_price_off_peak_fix', 'var_6m_price_peak_fix', 'var_6m_price_mid_peak_fix']
all_6m_columns = ['var_6m_price_off_peak', 'var_6m_price_peak', 'var_6m_price_mid_peak']

featureEng['avg_6m_price_var'] = featureEng[var_6m_columns].mean(axis=1)
featureEng['avg_6m_price_fix'] = featureEng[fix_6m_columns].mean(axis=1)
featureEng['avg_6m_price_all'] = featureEng[all_6m_columns].mean(axis=1)

# 2. Mean price changes across all types for yearly period
var_year_columns = ['var_year_price_off_peak_var', 'var_year_price_peak_var', 'var_year_price_mid_peak_var']
fix_year_columns = ['var_year_price_off_peak_fix', 'var_year_price_peak_fix', 'var_year_price_mid_peak_fix']
all_year_columns = ['var_year_price_off_peak', 'var_year_price_peak', 'var_year_price_mid_peak']

featureEng['avg_year_price_var'] = featureEng[var_year_columns].mean(axis=1)
featureEng['avg_year_price_fix'] = featureEng[fix_year_columns].mean(axis=1)
featureEng['avg_year_price_all'] = featureEng[all_year_columns].mean(axis=1)

# 3. Volatility measures - standard deviation and range of price changes
featureEng['volatility_6m_price_var'] = featureEng[var_6m_columns].std(axis=1)
featureEng['volatility_year_price_var'] = featureEng[var_year_columns].std(axis=1)
featureEng['volatility_6m_price_fix'] = featureEng[fix_6m_columns].std(axis=1)
featureEng['volatility_year_price_fix'] = featureEng[fix_year_columns].std(axis=1)

# Range (max - min) as volatility measure
featureEng['range_6m_price_var'] = featureEng[var_6m_columns].max(axis=1) - featureEng[var_6m_columns].min(axis=1)
featureEng['range_year_price_var'] = featureEng[var_year_columns].max(axis=1) - featureEng[var_year_columns].min(axis=1)

# 4. Ratios between different price periods
# Peak to off-peak ratios
featureEng['price_peak_to_offpeak_ratio_6m'] = featureEng['var_6m_price_peak'] / (featureEng['var_6m_price_off_peak'] + 0.001)
featureEng['price_peak_to_offpeak_ratio_year'] = featureEng['var_year_price_peak'] / (featureEng['var_year_price_off_peak'] + 0.001)

# Mid-peak to off-peak ratios
featureEng['price_midpeak_to_offpeak_ratio_6m'] = featureEng['var_6m_price_mid_peak'] / (featureEng['var_6m_price_off_peak'] + 0.001)
featureEng['price_midpeak_to_offpeak_ratio_year'] = featureEng['var_year_price_mid_peak'] / (featureEng['var_year_price_off_peak'] + 0.001)

# Peak to mid-peak ratios
featureEng['price_peak_to_midpeak_ratio_6m'] = featureEng['var_6m_price_peak'] / (featureEng['var_6m_price_mid_peak'] + 0.001)
featureEng['price_peak_to_midpeak_ratio_year'] = featureEng['var_year_price_peak'] / (featureEng['var_year_price_mid_peak'] + 0.001)

print("Basic price statistics and ratios created...")

Creating comprehensive price-related features...
Basic price statistics and ratios created...


In [28]:
# ENHANCED COMPREHENSIVE PRICE FEATURE ENGINEERING
print("Creating ENHANCED comprehensive price-related features...")

# 1. Mean price changes across all types for 6-month period
var_6m_columns = ['var_6m_price_off_peak_var', 'var_6m_price_peak_var', 'var_6m_price_mid_peak_var']
fix_6m_columns = ['var_6m_price_off_peak_fix', 'var_6m_price_peak_fix', 'var_6m_price_mid_peak_fix']
all_6m_columns = ['var_6m_price_off_peak', 'var_6m_price_peak', 'var_6m_price_mid_peak']

featureEng['avg_6m_price_var'] = featureEng[var_6m_columns].mean(axis=1)
featureEng['avg_6m_price_fix'] = featureEng[fix_6m_columns].mean(axis=1)
featureEng['avg_6m_price_all'] = featureEng[all_6m_columns].mean(axis=1)

# 2. Mean price changes across all types for yearly period
var_year_columns = ['var_year_price_off_peak_var', 'var_year_price_peak_var', 'var_year_price_mid_peak_var']
fix_year_columns = ['var_year_price_off_peak_fix', 'var_year_price_peak_fix', 'var_year_price_mid_peak_fix']
all_year_columns = ['var_year_price_off_peak', 'var_year_price_peak', 'var_year_price_mid_peak']

featureEng['avg_year_price_var'] = featureEng[var_year_columns].mean(axis=1)
featureEng['avg_year_price_fix'] = featureEng[fix_year_columns].mean(axis=1)
featureEng['avg_year_price_all'] = featureEng[all_year_columns].mean(axis=1)

# 3. Volatility measures - standard deviation and range of price changes
featureEng['volatility_6m_price_var'] = featureEng[var_6m_columns].std(axis=1)
featureEng['volatility_year_price_var'] = featureEng[var_year_columns].std(axis=1)
featureEng['volatility_6m_price_fix'] = featureEng[fix_6m_columns].std(axis=1)
featureEng['volatility_year_price_fix'] = featureEng[fix_year_columns].std(axis=1)

# Range (max - min) as volatility measure
featureEng['range_6m_price_var'] = featureEng[var_6m_columns].max(axis=1) - featureEng[var_6m_columns].min(axis=1)
featureEng['range_year_price_var'] = featureEng[var_year_columns].max(axis=1) - featureEng[var_year_columns].min(axis=1)

# 4. Ratios between different price periods
# Peak to off-peak ratios
featureEng['price_peak_to_offpeak_ratio_6m'] = featureEng['var_6m_price_peak'] / (featureEng['var_6m_price_off_peak'] + 0.001)
featureEng['price_peak_to_offpeak_ratio_year'] = featureEng['var_year_price_peak'] / (featureEng['var_year_price_off_peak'] + 0.001)

# Mid-peak to off-peak ratios
featureEng['price_midpeak_to_offpeak_ratio_6m'] = featureEng['var_6m_price_mid_peak'] / (featureEng['var_6m_price_off_peak'] + 0.001)
featureEng['price_midpeak_to_offpeak_ratio_year'] = featureEng['var_year_price_mid_peak'] / (featureEng['var_year_price_off_peak'] + 0.001)

# Peak to mid-peak ratios
featureEng['price_peak_to_midpeak_ratio_6m'] = featureEng['var_6m_price_peak'] / (featureEng['var_6m_price_mid_peak'] + 0.001)
featureEng['price_peak_to_midpeak_ratio_year'] = featureEng['var_year_price_peak'] / (featureEng['var_year_price_mid_peak'] + 0.001)

# 5. Fixed vs Variable price comparisons (ENHANCED)
print("Creating fixed vs variable price comparisons...")

for period in ['off_peak', 'peak', 'mid_peak']:
    # Fixed vs Variable differences
    featureEng[f'fix_var_diff_{period}_6m'] = featureEng[f'var_6m_price_{period}_fix'] - featureEng[f'var_6m_price_{period}_var']
    featureEng[f'fix_var_diff_{period}_year'] = featureEng[f'var_year_price_{period}_fix'] - featureEng[f'var_year_price_{period}_var']
    
    # Fixed vs Variable ratios
    featureEng[f'fix_to_var_ratio_{period}_6m'] = featureEng[f'var_6m_price_{period}_fix'] / (featureEng[f'var_6m_price_{period}_var'] + 0.001)
    featureEng[f'fix_to_var_ratio_{period}_year'] = featureEng[f'var_year_price_{period}_fix'] / (featureEng[f'var_year_price_{period}_var'] + 0.001)

# 6. Price consistency measures (NEW)
print("Creating price consistency features...")

# Coefficient of variation (relative volatility)
featureEng['cv_6m_price_var'] = featureEng['volatility_6m_price_var'] / (featureEng['avg_6m_price_var'].abs() + 0.001)
featureEng['cv_year_price_var'] = featureEng['volatility_year_price_var'] / (featureEng['avg_year_price_var'].abs() + 0.001)
featureEng['cv_6m_price_fix'] = featureEng['volatility_6m_price_fix'] / (featureEng['avg_6m_price_fix'].abs() + 0.001)
featureEng['cv_year_price_fix'] = featureEng['volatility_year_price_fix'] / (featureEng['avg_year_price_fix'].abs() + 0.001)

# Price consistency score
featureEng['price_consistency_score'] = 1 / (featureEng['cv_year_price_var'] + 0.001)

print("Enhanced price features created successfully!")

Creating ENHANCED comprehensive price-related features...
Creating fixed vs variable price comparisons...
Creating price consistency features...
Enhanced price features created successfully!


In [29]:
featureEng.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14606 entries, 0 to 14605
Columns: 103 entries, id to price_consistency_score
dtypes: datetime64[ns](4), float64(77), int32(6), int64(12), object(4)
memory usage: 11.1+ MB


In [30]:
from sklearn.feature_selection import VarianceThreshold

# Get numeric columns only (excluding target)
numeric_columns = featureEng.select_dtypes(include=[np.number]).columns.tolist()
if 'churn' in numeric_columns:
    numeric_columns.remove('churn')

# Find low variance features
feature_variances = featureEng[numeric_columns].var()
low_variance_features = feature_variances[feature_variances < 0.01].index.tolist()

print(f"Dropping {len(low_variance_features)} low variance features:")
for feature in low_variance_features:
    print(f"- {feature}: variance = {feature_variances[feature]:.6f}")

# Drop low variance features
featureEng = featureEng.drop(columns=low_variance_features)

print(f"\nDataset shape after dropping low variance features: {featureEng.shape}")

Dropping 24 low variance features:
- forecast_price_energy_off_peak: variance = 0.000606
- forecast_price_energy_peak: variance = 0.002405
- var_year_price_off_peak_var: variance = 0.000000
- var_year_price_peak_var: variance = 0.000000
- var_year_price_mid_peak_var: variance = 0.000000
- var_6m_price_off_peak_var: variance = 0.000000
- var_6m_price_peak_var: variance = 0.000000
- var_6m_price_mid_peak_var: variance = 0.000000
- contract_expired: variance = 0.000000
- dec_jan_diff_off_peak_var: variance = 0.000145
- dec_jan_diff_peak_var: variance = 0.000115
- dec_jan_diff_mid_peak_var: variance = 0.000049
- dec_jan_abs_diff_off_peak_var: variance = 0.000109
- dec_jan_abs_diff_peak_var: variance = 0.000106
- dec_jan_abs_diff_mid_peak_var: variance = 0.000047
- dec_jan_avg_diff_var: variance = 0.000048
- avg_6m_price_var: variance = 0.000000
- avg_year_price_var: variance = 0.000000
- volatility_6m_price_var: variance = 0.000000
- volatility_year_price_var: variance = 0.000000
- range_6